File: viz/src/gol/coarse.ts
// Coarse-graining: liveness grid -> placed objects (port of holons/coarse.py).
// A placed object is a Moore-connected component, canonicalised to a
// translation-invariant pattern key plus an anchor (min-corner).
import type { LifeField } from "./life";

export interface PlacedObject {
  patternKey: string; // canonical cells, sorted, as a string
  anchor: [number, number]; // (minRow, minCol)
  cells: Array<[number, number]>; // absolute cells of this component
  size: number; // |L|: live-cell count == structural support breadth
}

// Moore-connected (8-neighbour) components of live cells.
export function mooreComponents(f: LifeField): Array<Array<[number, number]>> {
  const visited = new Uint8Array(f.w * f.h);
  const comps: Array<Array<[number, number]>> = [];
  for (let r0 = 0; r0 < f.h; r0++) {
    for (let c0 = 0; c0 < f.w; c0++) {
      const idx0 = r0 * f.w + c0;
      if (f.cells[idx0] === 0 || visited[idx0]) continue;
      const comp: Array<[number, number]> = [];
      const stack: Array<[number, number]> = [[r0, c0]];
      while (stack.length > 0) {
        const [r, c] = stack.pop()!;
        if (r < 0 || r >= f.h || c < 0 || c >= f.w) continue;
        const idx = r * f.w + c;
        if (visited[idx] || f.cells[idx] === 0) continue;
        visited[idx] = 1;
        comp.push([r, c]);
        for (let dr = -1; dr <= 1; dr++)
          for (let dc = -1; dc <= 1; dc++)
            if (dr !== 0 || dc !== 0) stack.push([r + dr, c + dc]);
      }
      comps.push(comp);
    }
  }
  return comps;
}

export function canonicalise(cells: Array<[number, number]>): PlacedObject {
  let minR = Infinity;
  let minC = Infinity;
  for (const [r, c] of cells) {
    if (r < minR) minR = r;
    if (c < minC) minC = c;
  }
  const rel = cells
    .map(([r, c]): [number, number] => [r - minR, c - minC])
    .sort((a, b) => (a[0] - b[0]) || (a[1] - b[1]));
  const patternKey = rel.map(([r, c]) => `${r},${c}`).join(";");
  return { patternKey, anchor: [minR, minC], cells, size: cells.length };
}

export function coarseState(f: LifeField): PlacedObject[] {
  return mooreComponents(f).map(canonicalise);
}

// Centroid of a component (for display).
export function centroid(obj: PlacedObject): [number, number] {
  let sr = 0;
  let sc = 0;
  for (const [r, c] of obj.cells) {
    sr += r;
    sc += c;
  }
  return [sr / obj.cells.length, sc / obj.cells.length];
}

File: holons/coarse.py
"""Coarse-graining: liveness grid -> set of placed objects.

The catalogue `O` of §7 is all finite Life patterns up to translation,
represented as the canonical frozenset of `(row, col)` offsets with minimum
corner at `(0, 0)`. A placed object is a `(pattern, anchor)` pair.

The coarse-graining sends a fine state to the frozenset of placed objects
obtained from its Moore-connected (8-neighbour) live-cell components. The
clustering rule is defined for every fine state and lands in `O × Z^2`
by construction, so reaction products are catalogued rather than
exceptional (annihilation being the empty set).
"""
from __future__ import annotations

import numpy as np

from .patterns import Anchor, Pattern


PlacedObject = tuple[Pattern, Anchor]
CoarseState = frozenset[PlacedObject]


def moore_components(beta: np.ndarray) -> list[list[tuple[int, int]]]:
    """Return the Moore-connected components of live cells in `beta`.

    Each component is a list of `(row, col)` tuples. The order of
    components and of cells within a component is unspecified; downstream
    code must not depend on it.
    """
    visited = np.zeros(beta.shape, dtype=bool)
    components: list[list[tuple[int, int]]] = []
    rows, cols = beta.shape
    for r0 in range(rows):
        for c0 in range(cols):
            if beta[r0, c0] == 0 or visited[r0, c0]:
                continue
            component: list[tuple[int, int]] = []
            stack: list[tuple[int, int]] = [(r0, c0)]
            while stack:
                r, c = stack.pop()
                if not (0 <= r < rows and 0 <= c < cols):
                    continue
                if visited[r, c] or beta[r, c] == 0:
                    continue
                visited[r, c] = True
                component.append((r, c))
                for dr in (-1, 0, 1):
                    for dc in (-1, 0, 1):
                        if dr == 0 and dc == 0:
                            continue
                        stack.append((r + dr, c + dc))
            components.append(component)
    return components


def canonicalise(cells: list[tuple[int, int]]) -> PlacedObject:
    """Return `(canonical pattern, anchor)` for a non-empty list of cells.

    The anchor is `(min row, min col)` over the cells. The canonical
    pattern is the frozenset of cells shifted so that the anchor maps to
    `(0, 0)`. Two cell lists that are translates of each other produce
    the same canonical pattern.
    """
    if not cells:
        raise ValueError("cannot canonicalise empty component")
    min_r = min(r for r, _ in cells)
    min_c = min(c for _, c in cells)
    pattern: Pattern = frozenset(
        (r - min_r, c - min_c) for r, c in cells
    )
    return pattern, (min_r, min_c)


def coarse_state(beta: np.ndarray) -> CoarseState:
    """Coarse-grain a fine liveness grid to a frozenset of placed objects."""
    components = moore_components(beta)
    return frozenset(canonicalise(c) for c in components if c)
